HyperVExtension/src/HyperVExtension/Services/VMGalleryService.cs

```diff
@@ -0,0 +1,129 @@
+
+    private readonly IDownloaderService _downloaderService;
+
+    private readonly JsonSerializerOptions _jsonOptions = new()
+    {
+        PropertyNameCaseInsensitive = true,
+        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+        TypeInfoResolver = JsonSourceGenerationContext.Default,
+        AllowTrailingCommas = true,
+    };
+
+    public VMGalleryService(IDownloaderService downloaderService)
+    {
+        _downloaderService = downloaderService;
+    }
+
+    private static readonly Uri VmGalleryUrl = new("https://go.microsoft.com/fwlink/?linkid=851584");
+
+    private VMGalleryImageList _imageList = new();
+
+    /// <inheritdoc cref="IVMGalleryService.GetGalleryImagesAsync"/>
+    public async Task<VMGalleryImageList> GetGalleryImagesAsync()
+    {
+        // If we have already downloaded the image list before, return it.
+        if (_imageList.Images.Count > 0)
+        {
+            return _imageList;
+        }
+
+        var emptyList = new VMGalleryImageList();
+        try
+        {
+            var cancellationTokenSource = new CancellationTokenSource();
+
+            // This should be quick as the file is around 28.0 KB, so a 5 minute timeout should be ok on the worst network.
+            cancellationTokenSource.CancelAfter(TimeSpan.FromMinutes(5));
+
+            // get the JSON data
+            var resultJson = await _downloaderService.DownloadStringAsync(VmGalleryUrl.AbsoluteUri, cancellationTokenSource.Token);
+            _imageList = JsonSerializer.Deserialize(resultJson, typeof(VMGalleryImageList), _jsonOptions) as VMGalleryImageList ?? emptyList;
+
+            // Now we need to download the base64 images for the symbols (icons). So they can be used within an adaptive card.
+            foreach (var image in _imageList.Images)
+            {
+                if (!string.IsNullOrEmpty(image.Symbol.Uri))
+                {
+                    var byteArray = await _downloaderService.DownloadByteArrayAsync(image.Symbol.Uri, cancellationTokenSource.Token);
+                    if (!ValidateFileSha256Hash(byteArray, image.Symbol.Hash))
+                    {
+                        Logging.Logger()?.ReportError(ComponentName, $"Symbol Hash '{image.Symbol.Hash}' validation failed for image with name '{image}'. Symbol uri: '{image.Symbol}'");
+                        continue;
+                    }
+
+                    image.Symbol.Base64Image = Convert.ToBase64String(byteArray);
+                }
+            }
+        }
+        catch (Exception ex)
+        {
+            Logging.Logger()?.ReportError(ComponentName, $"Unable to retrieve VM gallery images", ex);
+        }
+
+        return _imageList;
+    }
+
+    /// <inheritdoc cref="IVMGalleryService.GetDownloadedArchiveFileName"/>
+    public string GetDownloadedArchiveFileName(VMGalleryImage image)
+    {
+        var hash = image.Disk.Hash;
+        var hexValues = hash.Split(':').Last();
```

Comment:
Are we sure splitting on `:` is always correct?  Is this character duplicated somewhere else , resulting in a chance that both characters can become desynched?

Reply:
Yes, here is an example of what this hash looks like:

```Json
"disk": {
        "uri": "https://download.microsoft.com/download/5/4/c/54cf292f-a9e3-47d3-9b4d-5659dc1c6189/MSIX Packaging Tool 10182019.zip",
        "hash": "sha256:09C50382D496C5DF2C96034EB69F20D456B25308B3F672257D55FD8202DDF84B",
        "archiveRelativePath": "127gb__Image-Evl_Win10_001.vhdx",
        },
```

We guarantee it to look like this. If it is changed in the future it would like be changed by our team, in which case we can update this at that time. But this hasn't changed since it was created in 2017 - 2018.